### PR TITLE
Quality engine and Q005

### LINKED ISSUE #188: Validation (2017): Q005 (Loan Amount)
**Edit #:** Q005
**Type:** Quality
**Category:** Loan Application Register (only)
**Column:** Loan Amount

If type of purchaser = 1-4, and property type = 1 or 2, then loan amount should be ≤ $1,203 ($1,203 thousand).

Additional resources:
* [FFIEC's 2016 HMDA File Spec](https://www.ffiec.gov/hmda/pdf/spec2016.pdf)
* [FFIEC's 2016 HMDA Edits](https://www.ffiec.gov/hmda/pdf/edit2016.pdf)
* [CFPB's 2017 HMDA File Spec](http://www.consumerfinance.gov/hmda/static/for-filers/2017/2017-HMDA-File-Specifications.pdf)

### LINKED ISSUE #372: Create Quality Engine
Similar to the `LarValidityEngine` and `LarSyntacticalEngine`, we need a `LarQualityEngine`.  Associated API calls should be modified and/or created, and tests are should be included.

### PR DESCRIPTION
**Pull request by nickgrippin**
Closes #372
Closes #188 

Any comments are greatly appreciated!

### CODE REVIEW

File: validation/src/test/scala/hmda/validation/rules/lar/quality/Q005Spec.scala [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import hmda.validation.rules.EditCheck
+import hmda.validation.rules.lar.{ BadValueUtils, LarEditCheckSpec }
+import org.scalacheck.Gen
+
+class Q005Spec extends LarEditCheckSpec with BadValueUtils {
+  property("Whenever purchaser type and property type are relevant, loan amount must be <= 1203") {
+    forAll(larGen) { lar =>
+      lar.mustPass
+    }
+  }
```
Comment (lgdean) [inline validation/src/test/scala/hmda/validation/rules/lar/quality/Q005Spec.scala]: I don't think the property text matches what is actually happening here.

File: validation/src/test/scala/hmda/validation/engine/lar/quality/LarQualityEngineSpec.scala [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    with LarGenerators
+    with LarQualityEngine {
+
+  property("A LAR must pass quality checks") {
+    forAll(larGen) { lar =>
+      whenever(lar.id == 2) {
+        checkQuality(lar).isSuccess mustBe true
+      }
+    }
+  }
```
Comment (lgdean) [inline validation/src/test/scala/hmda/validation/engine/lar/quality/LarQualityEngineSpec.scala]: I'm a little skeptical of this spec, but no more than the ones it's based on. I'm not convinced that all our generated LARs will pass quality checks. That said, if it's currently passing, then I wouldn't let it stand in the way of merging. Probably.

File: validation/src/test/scala/hmda/validation/rules/lar/quality/Q005Spec.scala [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import hmda.validation.rules.EditCheck
+import hmda.validation.rules.lar.{ BadValueUtils, LarEditCheckSpec }
+import org.scalacheck.Gen
+
+class Q005Spec extends LarEditCheckSpec with BadValueUtils {
+  property("All generated lars must pass") {
+    forAll(larGen) { lar =>
+      lar.mustPass
+    }
+  }
```
Comment (lgdean) [inline validation/src/test/scala/hmda/validation/rules/lar/quality/Q005Spec.scala]: I'm confused. Couldn't `larGen` in theory give you a LAR with a purchaserType of 3, a loan.propertyType of 2, and a loan.amount of 1200?

File: validation/src/test/scala/hmda/validation/engine/lar/quality/LarQualityEngineSpec.scala (hunk 2 above)
Comment (nickgrippin) [inline validation/src/test/scala/hmda/validation/engine/lar/quality/LarQualityEngineSpec.scala]: I think I agree with you.  Previously, I know that I've had to change the `LarGenerators` to match our tests.  I imagine that's what we'll continue to do: whenever we find a quality check that will make the generators fail, we can just modify the generators.

File: api/src/main/scala/hmda/api/processing/lar/SingleLarValidation.scala [hunk 4]
```diff
@@ -28,6 +29,9 @@ class SingleLarValidation extends Actor with ActorLogging with LarEngine {
     case CheckValidity(lar) =>
       log.debug(s"Checking validity on LAR: ${lar.toCSV}")
       sender() ! validationErrors(lar, checkValidity)
+    case CheckQuality(lar) =>
+      log.debug(s"Checking quality on LAR: ${lar.toCSV}")
+      sender() ! validationErrors(lar, checkQuality)
```
Comment (lgdean) [inline api/src/main/scala/hmda/api/processing/lar/SingleLarValidation.scala]: I can replace this line with `sender() ! Nil` and all the tests still pass. I'd prefer if we cover that case. Not necessarily with an integration test, but with something.

File: validation/src/test/scala/hmda/validation/engine/lar/quality/LarQualityEngineSpec.scala (hunk 2 above)
Comment (lgdean) [inline validation/src/test/scala/hmda/validation/engine/lar/quality/LarQualityEngineSpec.scala]: I think that approach will make the generators overly complex. (I also think this question has come up before, but I don't have time to dig up the thread right now.)

File: validation/src/test/scala/hmda/validation/rules/lar/quality/Q005Spec.scala (hunk 3 above)
Comment (nickgrippin) [inline validation/src/test/scala/hmda/validation/rules/lar/quality/Q005Spec.scala]: `loan.amount` is created with `Gen.posNum[Int]`, which in all the tests I've seen, never generates a number higher than 99.  I don't know if that's behavior that we've defined.

File: api/src/main/scala/hmda/api/processing/lar/SingleLarValidation.scala (hunk 4 above)
Comment (nickgrippin) [inline api/src/main/scala/hmda/api/processing/lar/SingleLarValidation.scala]: This should be fixed with the newest commit

File: validation/src/test/scala/hmda/validation/engine/lar/quality/LarQualityEngineSpec.scala (hunk 2 above)
Comment (lgdean) [inline validation/src/test/scala/hmda/validation/engine/lar/quality/LarQualityEngineSpec.scala]: The most relevant bits I found in a search are in #82, specifically [here](https://github.com/cfpb/hmda-platform/pull/82#discussion_r54579326). I think we've also discussed it since then in the context of particular edits, but perhaps not in github issues (or not that I found).

File: validation/src/main/scala/hmda/validation/rules/lar/quality/Q005.scala [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+import hmda.validation.rules.EditCheck
+import hmda.validation.dsl.PredicateCommon._
+import hmda.validation.dsl.PredicateSyntax._
+
+object Q005 extends EditCheck[LoanApplicationRegister] {
+  override def apply(lar: LoanApplicationRegister): Result = {
+    when((lar.purchaserType is containedIn(List(1, 2, 3, 4))) and
+      (lar.loan.propertyType is containedIn(List(1, 2)))) {
+      lar.loan.amount is lessThanOrEqual(1203)
+    }
```
Comment (jmarin) [inline validation/src/main/scala/hmda/validation/rules/lar/quality/Q005.scala]: These values (i.e. 1203) are probably going to change through time (they have in the last few years). We need a mechanism to make that configurable. I suggest adding this to `application.conf` in the validation project with a setting, and providing the ability to override the value with an environment variable. Something like this:

`hmda.validation.quality.Q005.loan.amount=1203`
`hmda.validation.quality.Q005.loan.amount=${?Q005_LOAN_AMOUNT}`

You can then read the value by doing something like:

```scala
val config = ConfigFactory.load()
val loanAmount = config.getInt("hmda.validation.quality.Q005.loan.amount")
```

File: validation/src/test/scala/hmda/validation/engine/lar/quality/LarQualityEngineSpec.scala [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    with LarGenerators
+    with LarQualityEngine {
+
+  property("A LAR must pass quality checks") {
+    val lines = Source.fromFile(new File("parser/src/test/resources/txt/FirstTestBankData_clean_407_2017.txt")).getLines()
+    val lars = lines.drop(1).map(l => LarCsvParser(l))
+
+    lars.foreach { lar =>
+      checkQuality(lar).isSuccess mustBe true
+    }
```
Comment (jmarin) [inline validation/src/test/scala/hmda/validation/engine/lar/quality/LarQualityEngineSpec.scala]: Just a warning here, these files were checked for syntactical and validity edits only. They may fail some quality and macro edits.

File: validation/src/test/scala/hmda/validation/rules/lar/quality/Q005Spec.scala [hunk 7]
```diff
@@ -0,0 +1,45 @@
+package hmda.validation.rules.lar.quality
+
+import hmda.model.fi.lar.LoanApplicationRegister
+import hmda.validation.rules.EditCheck
+import hmda.validation.rules.lar.{ BadValueUtils, LarEditCheckSpec }
+import org.scalacheck.Gen
+
+class Q005Spec extends LarEditCheckSpec with BadValueUtils {
+  property("All lars with loan amounts less than $1203 must pass") {
+    forAll(larGen, Gen.choose(0, 1203)) { (lar, x) =>
```
Comment (jmarin) [inline validation/src/test/scala/hmda/validation/rules/lar/quality/Q005Spec.scala]: Same comment as above regarding the hardcoded $ amount

File: validation/src/test/scala/hmda/validation/rules/lar/quality/Q005Spec.scala [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  property("Whenever property type is irrelevant, lar must pass") {
+    forAll(larGen, irrelevantProperty) { (lar, x) =>
+      val newLoan = lar.loan.copy(propertyType = x)
+      val newLar = lar.copy(loan = newLoan)
+      newLar.mustPass
+    }
+  }
+
+  property("A lar with relevant property and purchaser types with a loan amount > 1203 must fail") {
+    forAll(larGen, Gen.choose(1, 4), Gen.oneOf(1, 2), Gen.choose(1204, Int.MaxValue)) { (lar, x, y, amount) =>
```
Comment (jmarin) [inline validation/src/test/scala/hmda/validation/rules/lar/quality/Q005Spec.scala]: Same comment as above regarding hardcoded $ amount

File: validation/src/test/scala/hmda/validation/engine/lar/quality/LarQualityEngineSpec.scala (hunk 6 above)
Comment (nickgrippin) [inline validation/src/test/scala/hmda/validation/engine/lar/quality/LarQualityEngineSpec.scala]: Should we modify them if we find errors in future quality/macro edits, or would you suggest a different testing approach?

The tests are currently passing, I was trying a new approach that didn't rely on the generators.

Comment (jmarin) [inline validation/src/test/scala/hmda/validation/engine/lar/quality/LarQualityEngineSpec.scala]: In general, a "perfect" file for quality/macro is rare. Many of these edits won't pass for a HMDA filing, but that doesn't mean the file submission is wrong. Users need to provide reasons and/or confirm that their data is correct. This is outside of the scope of these edits though. Regarding using these files for quality/macro checks, that's a question for @debseidner

Comment (debseidner) [inline validation/src/test/scala/hmda/validation/engine/lar/quality/LarQualityEngineSpec.scala]: @jmarin is correct. Most filings will have quality and macro quality errors. I cannot separate out quality and macro so these stats have both but here are the numbers of filings with quality/macro edits.

-  2012: 92% of institutions have quality/macro errors, 19% of all LARs
-  2013: 92% of institutions have quality/macro errors, 18% of all LARs
-  2014: 91% of institutions have quality/macro errors, 16% of all LARs
-  2015: 93% of institutions have quality/macro errors, 17% of all LARs

File: validation/src/test/scala/hmda/validation/engine/lar/quality/LarQualityEngineSpec.scala (hunk 2 above)
Comment (lgdean) [inline validation/src/test/scala/hmda/validation/engine/lar/quality/LarQualityEngineSpec.scala]: For future reference, #271 was the discussion I had in mind. (Thanks to @schbetsy for finding it! Turns out I was searching only PRs, but it wasn't a PR.)

Comment (nickgrippin) [inline validation/src/test/scala/hmda/validation/engine/lar/quality/LarQualityEngineSpec.scala]: Ok, that looks like a good summary.  Thanks!

File: api/src/test/scala/hmda/api/http/LarHttpApiSpec.scala [hunk 9]
```diff
@@ -51,22 +51,27 @@ class LarHttpApiSpec extends WordSpec with MustMatchers with ScalatestRouteTest
...
-      val badLar = lar.copy(agencyCode = 0, loan = badLoanType)
-      Post("/lar/validate", badLar) ~> larRoutes ~> check {
-        status mustEqual StatusCodes.OK
-        responseAs[List[ValidationError]].length mustBe 2
-      }
-      Post("/lar/validate?check=syntactical", badLar) ~> larRoutes ~> check {
-        status mustEqual StatusCodes.OK
-        responseAs[List[ValidationError]].length mustBe 1
+    "filter syntactical, validity, or quality only for invalid LAR with " +
+      "both syntactical, validity, and quality errors" in {
```
Comment (lgdean) [inline api/src/test/scala/hmda/api/http/LarHttpApiSpec.scala]: this is a nitpick, but "both" doesn't quite work with more than 2 things. maybe "all 3 kinds of errors"?

**Comment by schbetsy** [comment]
LGTM

**Comment by jmarin** [comment]
LGTM
